### ISSUE TITLE
Extend the logic symbols

### ISSUE THREAD

**Issue by leoheck** [drom/wavedrom#115]
It is possible to extend the symbols including:
- A generic symbol (could be a rectangle) with N inputs and N outputs? With customizable internal label? In this way, the user can create its own symbols. 
- A way to user add new symbols.

**Comment by leoheck** [comment]
It would be also nice to give support for multiples inputs on the standard symbols like in those you are using. For example with a NAND with 2, 3, 4, 5 inputs.. 

And the possibility to create complex gates like AOI (and-or-inv, without wires between the gates)

![250px-aoi22symbol svg](https://cloud.githubusercontent.com/assets/1277920/8508748/a9700214-2253-11e5-8723-77caf8242c4d.png)

**Comment by drom** [comment]
All good points.
I like the idea with complex gates. Do you have and idea how they may be described?

Now the following descripion:
```js
{ assign:[
  ["out",
    ["~|",
      ["&", "a", "b"],
      ["&", "c", "d"]
    ]
  ]
]}
```
Will generate:
![wavedrom](https://cloud.githubusercontent.com/assets/511872/8603801/53a796be-262f-11e5-932d-e4a0dcbaa0b7.png)

Do you think that this should produce complex gate as is, or different description can be given?

**Comment by leoheck** [comment]
Awesome!

I don't know, we could add some keyword indicating a complex gate. But I think you can propose something better since you know the project better than me.

```
{ assign:[
  ["out",
    ["~|c",   <------- look the c here
      ["&", "a", "b"],
      ["&", "c", "d"]
    ]
  ]
]}
```

![53a796be-262f-11e5-932d-e4a0dcbaa0b7](https://cloud.githubusercontent.com/assets/1277920/8603948/d8a61f94-2651-11e5-843f-a641445fdd06.png)

**Comment by leoheck** [comment]
It will be nice to have symbols with multiple outputs.. for example a flip-flop with Q and ~Q output.

**Comment by drom** [comment]
How would you describe multi-output gate?

**Comment by leoheck** [comment]
I can't find a good way by now but I am adding some symbols here to consider on the thinking process.

![toggle_symbol](https://cloud.githubusercontent.com/assets/1277920/8606621/3209e0a0-2663-11e5-8490-e230e4bc00d1.png)

- A toggle and a mutex generic box components with a user-defined labels and 2 outputs.
- A Muller C-element (could also be a generic circle symbol) with 3 inputs.

I will add some ideas here when I had an idea.

**Comment by drom** [comment]
We should probably separate the notion of element shape, number of pins and label.
That will require clear notation and present the challenge. Any thoughts are welcome.

**Comment by YPetremann** [comment]
I think a definition system could be great, we define a gate model by it's way to be rendered, it's shape, label and ports (it could have multiple input and output port), then we instantiate the model how many time we want, finally we tell to connect edge to ports. edges could have style and label too. A port could accept many edges

'''
{
  models:{
    toggle:{
      label:"toggle",
      shape:"rect",
      ports:[
        {mode:"input",side:"left",default:true},
        {mode:"output",side:"right",name:"q"},
        {mode:"output",side:"top",name:"nq", style:"inverse"}
      ]
    }
  },
  blocks:{
    ["toggle","toggle"]
  },
  assign:[
    ["a","toggle"]
    ["toggle:q","b"]
    ["toggle:nq","c"]
  ]}
'''
